threadloop/threadloop.py

```diff
@@ -124,7 +125,9 @@ def on_done(tornado_future):
             future.set_exception(exception)
 
         self._io_loop.add_callback(
-            lambda: fn(*args, **kwargs).add_done_callback(on_done)
+            lambda: gen.maybe_future(
+                fn(*args, **kwargs)
```

Comment:
This will barf if `fn` is not async and raises an exception. You basically have to handle both cases.